File: 03-hard/0006-longest-common-subsequence/0006-longest-common-subsequence.cpp
/*
  Copyright (C) 2015 Daniel Rodriguez

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// Headers for test case
#include <fstream>
#include <iostream>
#include <string>

// Headers for the implementation
#include <algorithm>
#include <future>
#include <iterator>
#include <limits>
#include <memory>
#include <type_traits>
#include <vector>

/*
Hirschberg algorithm:
  - http://dl.acm.org/citation.cfm?id=360861
  - http://bioinfo.ict.ac.cn/~dbu/AlgorithmCourses/Lectures/Hirschberg1975.pdf
*/

// C++ 14 autodetects return type from algo_c
// GCC 4.9.2: compile with -std=c++14 (c++1y is deprecated but works)
// Uncomment trailing return types below if C++14 is unavailable

const size_t maxsize_t = std::numeric_limits<size_t>::max();


template<typename iter>
auto
algo_b(iter a1, iter a2, iter b1, iter b2, size_t kmax=maxsize_t)
    // -> std::vector<size_t>
{
    std::vector<size_t> k1(std::distance(b1, b2) + 1, 0);

    for(auto a=a1; a != a2; ++a) {
        auto aelem = *a;

        size_t k0j1 = 0;
        size_t k1j1 = 0;
        auto j = 1;
        for(auto b=b1; b != b2; b++, j++) {
            auto k0j = k1[j];
            k1[j] = k1j1 = (aelem == *b) ? k0j1 + 1 : std::max(k1j1, k0j);
            if(k1j1 >= kmax)
                return k1;
            k0j1 = k0j;
        }
    }
    return k1;
}

template<typename iter, typename iterout>
auto
algo_c2(iter a1, iter a2, iter b1, iter b2, iterout v, size_t lmax, bool parallel=false)
    // -> void
{
    auto m = std::distance(a1, a2);
    auto n = std::distance(b1, b2);

    if(not m or not n)
        return;  // one (or both) of the sequences are empty

    // Optimize out common prefixes
    auto abmis = std::mismatch(a1, std::next(a1, std::min(m, n)), b1);
    if(a1 != abmis.first) {
        std::copy(a1, abmis.first, v);
        a1 = abmis.first;
        b1 = abmis.second;
    }

    if(b1 == b2 or a1 == a2)
        return;  // After optimization one (or both) of the sequences are empty

    // Common Suffix optimization
    auto suffix_len = 0;
    auto a2rs = std::reverse_iterator<decltype(a2)>(a2);
    auto b2rs = std::reverse_iterator<decltype(b2)>(b2);

    m = std::distance(a1, a2);  // a1 may have changed above
    n = std::distance(b1, b2);
    auto abrmis = std::mismatch(a2rs, std::next(a2rs, std::min(m, n)), b2rs);
    if(a2rs != abrmis.first) {
        suffix_len = std::distance(a2rs, abrmis.first);
        a2 = std::prev(a2, suffix_len);
        b2 = std::prev(b2, suffix_len);
    }

    if(b1 == b2 or a1 == a2) {
        if(suffix_len)  // suffix was optimized away
            std::copy(a2, std::next(a2, suffix_len), v);
        return;  // After optimization one (or both) of the sequences are empty
    }

    m = std::distance(a1, a2);  // a1 may have changed above
    if(m == 1) {
        // Trivial case. Longest subsequence can be 1 and must be a1 if any
        if(std::find(b1, b2, *a1) != b2)
            *(v++) = *a1;  // ++ is no-op in back_insert, but iterout may be anything

        if(suffix_len)  // suffix was optimized away
            std::copy(a2, std::next(a2, suffix_len), v);

        return;
    }

    // Partition a
    auto ai = std::next(a1, m / 2);

    // Retrieve LCS lengths for forward part of a
    auto l1 = algo_b(a1, ai, b1, b2, lmax);  // will be used as output of l1 + l2

    // Prepare reversal of 2nd part of a and complete b
    auto air = std::reverse_iterator<decltype(ai)>(ai);
    auto a2r = std::reverse_iterator<decltype(a2)>(a2);

    auto b1r = std::reverse_iterator<decltype(b1)>(b1);
    auto b2r = std::reverse_iterator<decltype(b2)>(b2);

    // Retrieve LCS lengths for back part of a
    const auto &l2 = algo_b(a2r, air, b2r, b1r, lmax);

    // find best patition point k for b
    // sum l1 + l2_rev onto l1 (actually l1_r + l2 to get k in 1 step)
    auto l1rbegin = l1.rbegin();
    std::transform(l1rbegin, l1.rend(), l2.begin(), l1rbegin, std::plus<size_t>());

    // lowest pos with max sum - max yields right iter ... reversed sum above
    auto l1begin = l1.begin();
    auto k = std::distance(l1begin, std::max_element(l1begin, l1.end()));
    // Partition b
    auto bk = std::next(b1, k);

    // Limit the maximum sum to the already known max
    lmax = l1[k];


    // Solve the smaller problems
#ifdef _GLIBCXX_HAS_GTHREADS
    if(parallel) {
#if 0
        std::vector<typename std::decay<decltype(*a1)>::type> vasync1;
        auto v1 = std::back_inserter(vasync1);
        auto fut1 = std::async(std::launch::async, algo_c2<iter, iterout>,
                               a1, ai, b1, bk, v1, lmax, parallel);
#else
        // Partitions: a1 -> ai & b1 -> bk
        algo_c2(a1, ai, b1, bk, v, lmax, parallel);
#endif
        std::vector<typename std::decay<decltype(*a1)>::type> vasync2;
        auto v2 = std::back_inserter(vasync2);
        // Partitions: a1 -> a1 & b1 -> bk
        auto fut2 = std::async(std::launch::async, algo_c2<iter, iterout>,
                              ai, a2, bk, b2, v2, lmax, parallel);
#if 0
        fut1.get();
        std::copy(vasync1.begin(), vasync1.end(), v);
#endif
        fut2.get();
        std::copy(vasync2.begin(), vasync2.end(), v);
    } else
#else
    {
        // Partitions: a1 -> ai & b1 -> bk
        algo_c2(a1, ai, b1, bk, v, lmax, parallel);
        // Partitions: a1 -> a1 & b1 -> bk
        algo_c2(ai, a2, bk, b2, v, lmax, parallel);
    }
#endif

    // A suffix may have been saved ... add it
    if(suffix_len)
        std::copy(a2, std::next(a2, suffix_len), v);

    return;
}


// Overload to avoid having to pass a vector
template<typename iter>
auto
algo_c(iter a1, iter a2, iter b1, iter b2, bool parallel=true)
//    -> std::vector<typename std::decay<decltype(*a1)>::type>
{
    using itertype = typename std::decay<decltype(*a1)>::type;

    std::vector<itertype> v;
    algo_c2(a1, a2, b1, b2, std::back_inserter(v), maxsize_t, parallel);
    return v;
}


template <typename T>
auto
compute_lcs_vec(const T &a, const T &b, bool parallel=false)
//    -> std::vector<typename std::decay<decltype(*a.begin())>::type>
{
    return algo_c(a.begin(), a.end(), b.begin(), b.end(), parallel);
}


template <typename T>
auto
compute_lcs(const T &a, const T &b, bool parallel=false) // -> T
{
    auto v = compute_lcs_vec(a, b, parallel);
    return T(v.begin(), v.end());
}


int main(int argc, char *argv[]) {
    std::ifstream stream(argv[1]);
    std::string line;

    while(std::getline(stream, line)) {
        auto lbegin = line.begin();
        auto lend = line.end();
        auto sep = std::find(lbegin, lend, ';');

        const auto &v = algo_c(lbegin, sep, std::next(sep), lend, false);
        for(auto c: v)
            std::cout << c;
        std::cout << std::endl;
    }
    return 0;
}

File: 03-hard/0006-longest-common-subsequence/0006-longest-common-subsequence.py
#!/usr/bin/env python
# -*- coding: utf-8; py-indent-offset:4 -*-
###############################################################################
#
# Copyright (C) 2015 Daniel Rodriguez
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################
from __future__ import (absolute_import, division, print_function,)
#                        unicode_literals)

from bisect import bisect_left
import collections
import itertools
import sys


# Small Py2/3 compatibility layer
if sys.version_info.major == 2:
    range = xrange
    map = itertools.imap
else:
    pass


def compute_lcs_with_sets(t1, t2):
    '''
    Computes LCS (Longest Common Sequence) of strings t1, t2

      - Calculates the combinations for each of the strings for a given length

        The length starts at the shortest of the length of both strings

      - Those combinations are put into sets

      - An intersection of the sets is done. If any element is common to both
        the intersection will have at least 1 element (nothing prevents the
        strings from having 2 LCS at the same length)
    '''
    minlen = min(len(t1), len(t2))
    for l in range(minlen, 0, -1):
        # Calculate the combinations
        c1 = itertools.combinations(t1, l)
        c2 = itertools.combinations(t2, l)

        # Turn into sets
        s1 = set(c1)
        # print(s1)
        s2 = set(c2)
        # print(s2)

        # intersection
        lcs = s1 & s2

        if lcs:
            # At least 1 element - return a string
            # return ''.join(lcs.pop())
            return lcs.pop()

    # Out of the loop ... no LCS found
    return ''


def compute_lcs_with_product(t1, t2):
    '''
    Computes LCS (Longest Common Sequence) of strings t1, t2

      - Calculates the combinations for each of the strings for a given length

        The length starts at the shortest of the length of both strings

      - Those combinations are cartesian multiplied

      - If for any of the (x, y) cartesian pais x == y, we have a match
    '''
    minlen = min(len(t1), len(t2))
    for l in range(minlen, 0, -1):
        # Calculate the combinations
        c1 = itertools.combinations(t1, l)
        c2 = itertools.combinations(t2, l)

        p = itertools.product(c1, c2)
        for x, y in p:
            if x == y:
                return ''.join(x)

    # Out of the loop ... no LCS found
    return ''


MakePair = collections.namedtuple('Pair', ['index1', 'index2', 'nextpair'])


def Pairs(matches):
    traces = list()
    threshold = list()

    for index1, match in enumerate(matches):
        for index2 in match:
            index3 = bisect_left(threshold, index2)

            if index3 == len(threshold):
                threshold.append(index2)
                prefix = traces[index3 - 1] if index3 > 0 else None
                traces.append(MakePair(index1, index2, prefix))

            else:
                threshold[index3] = index2
                prefix = traces[index3 - 1] if index3 > 0 else None
                traces[index3] = MakePair(index1, index2, prefix)

    last = traces[-1] if len(traces) > 0 else None
    pair = last
    return len(threshold), pair


def Select(pair, s1, s2):
    string = []
    # pair is tuple (index1, index2, nextpair)
    while pair is not None:
        c = s2[pair.index2]
        string.append(c)
        pair = pair.nextpair

    return string


def Match(s1, s2):
    char2index = collections.defaultdict(collections.deque)
    for index, char in enumerate(s2):
        char2index[char].appendleft(index)

    matches = list(map(lambda c: char2index[c], s1))
    return matches


def Correspondence(s1, s2):
    matches = Match(s1, s2)
    length, last = Pairs(matches)
    return Select(last, s1, s2)


if __name__ == '__main__':
    test_cases = open(sys.argv[1], 'r')
    for test in test_cases:
        tstrip = test.rstrip('\r\n')
        if not tstrip:
            continue

        t1, t2 = test.split(';')
        lcs = Correspondence(t1, t2)

        print(''.join(reversed(lcs)))

    test_cases.close()
